shared/dataflow/codeql/dataflow/internal/DataFlowImplCommon.qll

```diff
@@ -958,6 +961,377 @@ module MakeImplCommon<InputSig Lang> {
+        trackedArgType(arg) and
+        relevantCallEdge(_, _, arg, p) and
+        trackedParamTypeCand(p)
+      )
+    }
+
+    /**
+     * Holds if `arg` is part of a value-propagating call path where the
+     * end-points have stronger types than the intermediate parameter and
+     * argument nodes.
+     */
+    private predicate trackedArgType(ArgNode arg) {
+      exists(ParamNode p |
+        trackedParamType(p) and
+        paramMustFlow(p, arg) and
+        trackedArgTypeCand(arg)
+      )
+    }
+
+    pragma[nomagic]
+    private predicate returnCallDeterminesParam(DataFlowCall call, ParamNode p) {
+      exists(ArgNode arg |
+        trackedArgType(arg) and
+        arg.argumentOf(call, _) and
+        paramMustFlow(p, arg)
+      )
+    }
+
+    private predicate returnCallLeavesParamUndetermined(DataFlowCall call, ParamNode p) {
+      trackedParamType(p) and
+      call.getEnclosingCallable() = getNodeEnclosingCallable(p) and
+      not returnCallDeterminesParam(call, p)
+    }
+
+    pragma[nomagic]
+    private predicate trackedParamWithType(ParamNode p, DataFlowType t, DataFlowCallable c) {
+      trackedParamType(p) and
+      c = getNodeEnclosingCallable(p) and
+      nodeDataFlowType(p, t)
+    }
+
+    pragma[nomagic]
+    private predicate dataFlowTakenCallEdgeIn(
+      DataFlowCall call, DataFlowCallable c, ArgNode arg, ParamNode p, boolean cc
+    ) {
+      Input::dataFlowTakenCallEdgeIn(call, c, cc) and
+      callEdge(call, c, arg, p) and
+      trackedParamType(p)
+    }
+
+    pragma[nomagic]
+    private predicate dataFlowTakenCallEdgeOut(
+      DataFlowCall call, DataFlowCallable c, ArgNode arg, ParamNode p
+    ) {
+      Input::dataFlowTakenCallEdgeOut(call, c) and
+      callEdge(call, c, arg, p) and
+      trackedArgType(arg) and
+      paramMustFlow(_, arg)
+    }
+
+    /**
+     * Holds if `t` is a possible type for an argument reaching the tracked
+     * parameter `p` through an in-going edge in the current data flow stage.
+     */
+    pragma[nomagic]
+    private predicate typeFlowParamTypeCand(ParamNode p, DataFlowType t) {
+      exists(ArgNode arg, boolean outercc |
+        dataFlowTakenCallEdgeIn(_, _, arg, p, outercc) and
+        if trackedArgType(arg) then typeFlowArgType(arg, t, outercc) else nodeDataFlowType(arg, t)
+      )
+    }
+
+    /**
+     * Holds if `t` is a possible type for the tracked parameter `p` in the call
+     * context `cc` and that the current data flow stage has reached this
+     * context.
+     */
+    private predicate typeFlowParamType(ParamNode p, DataFlowType t, boolean cc) {
+      exists(DataFlowCallable c |
+        Input::dataFlowNonCallEntry(c, cc) and
+        trackedParamWithType(p, t, c)
+      )
+      or
+      exists(DataFlowType t1, DataFlowType t2 |
+        cc = true and
+        typeFlowParamTypeCand(p, t1) and
+        nodeDataFlowType(p, t2) and
+        if typeStrongerThan(t2, t1) then t = t2 else (compatibleTypes(t1, t2) and t = t1)
```

Comment:
This pattern is repeated a lot, so perhaps but into a separate `bindingset`-annotated helper predicate.

Reply:
👍